performance/metrics.go

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package performance
+
+import (
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+const (
+	metricsNamespace     = "loadtest"
+	metricsSubSystemHTTP = "http"
+	metricsSubSystemWS   = "websocket"
+)
+
+type UserEntityMetrics struct {
+	HTTPRequestTimes     prometheus.Histogram
+	HTTPErrors           *prometheus.CounterVec
+	HTTPTimeouts         prometheus.Counter
+	WebSocketConnections prometheus.Gauge
+}
+
+type Metrics struct {
+	registry  *prometheus.Registry
+	ueMetrics UserEntityMetrics
+}
+
+func NewMetrics() *Metrics {
+	var m Metrics
+	m.registry = prometheus.NewRegistry()
+
+	m.ueMetrics.HTTPRequestTimes = prometheus.NewHistogram(prometheus.HistogramOpts{
```

Comment:
This is a great metric. Very useful when the client and the server are in different parts of the world. I think we should switch to tracking this instead of the server side API time, as this is what the user will actually feel. This will better react to network induced latencies or nginx issues.

Reply:
Looking at the graphs the averages for both client/server are practically identical (no surprise) but when you go into high percentile territory things start to differ.

I think both metrics are interesting on their own. The client response time might give you a better sense for the performance of the whole system but it could also be affected by more factors that don't have anything to do with the server, which is what we are ultimately testing.

Reply:
Yes, both are useful in their own context. I was thinking that the client metrics would be useful to monitor some end-to-end optimizations. For example, for a network delay of 10 ms, what effect does cutting down network transfer by 10% make. All those type of questions.